### ISSUE TITLE
Extricate LocalBarrier from load balancing code

### ISSUE THREAD

**Issue by rbuch** [UIUC-PPL/charm#2779]
The machinery underlying the `LocalBarrier` (i.e. the stuff for `AtSync` and invoking LB, etc.) is currently somewhat intertwined with the load balancing code. There's no reason this should be the case, instead the load balancer should merely be a registered user of a fully separated implementation that handles synchronization points, the services that should be run at those points, and resumption of the program.

**Comment by epmikida** [comment]
One point to bring up that was not really discussed in core is the notion of non-blocking AtSyncs. What I mean by this is that when using something like MetaBalancer, the LB framework can resume chares inline in the AtSync call instead of having them block on the AtSync barrier. Not sure how this would be handled in a generic framework, but its something that shouldn't be forgotten.

I ran into a similar thing with the LiveViz work when I was trying to use AtSync. Calling AtSync frequently caused pretty high overhead, so I chose to automatically resume chares in a lot of cases rather than waiting on full barrier. This was useful, for example, to grab a PEs load statistics once the PE itself had fully reached AtSync, but then immediately resume it rather than waiting on other PEs to all reach.

**Comment by rbuch** [comment]
> One point to bring up that was not really discussed in core is the notion of non-blocking AtSyncs. What I mean by this is that when using something like MetaBalancer, the LB framework can resume chares inline in the AtSync call instead of having them block on the AtSync barrier. Not sure how this would be handled in a generic framework, but its something that shouldn't be forgotten.
> 
> I ran into a similar thing with the LiveViz work when I was trying to use AtSync. Calling AtSync frequently caused pretty high overhead, so I chose to automatically resume chares in a lot of cases rather than waiting on full barrier. This was useful, for example, to grab a PEs load statistics once the PE itself had fully reached AtSync, but then immediately resume it rather than waiting on other PEs to all reach.

Yeah, that's a good point. One potential solution is if receivers have some way of being "global" or not. When there are "global" receivers, it becomes a real barrier, otherwise, they can just run whatever is to be run per client and then resume them.

In any case, whatever we do should certainly support the use case you're describing.

**Comment by rbuch** [comment]
To pick this back up again, to your earlier point @epmikida, a lot of receivers (like LB), do their own internal barriers e.g. in sending stats. So I think that's how this should be done, the barrier should just invoke its receivers when everything on its PE has called AtSync, and then those can do some synchronization internally should they need it.

**Comment by rbuch** [comment]
An (mostly complete) implementation of this is at #2955.
